### DISCUSSION TITLE
blog/inside-react-query

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#89] [Announcements]
# blog/inside-react-query

Taking a look under the hood of React Query

https://tkdodo.eu/blog/inside-react-query

**Comment by devagrawal09** [comment]
Great write-up! I was looking for something exactly like this

**Comment by yuvalshimoni** [comment]
Hi great blog, thanks! 
Small side question, why do you using 'useState' here instead of 'use useMemo'?
```js
const [observer] = React.useState(
    () =>
      new Observer<TQueryFnData, TError, TData, TQueryData, TQueryKey>(
        queryClient,
        defaultedOptions,
      ),
  )
```

**Reply by ivanjeremic** [reply]
I think since core handles caching there is no reason to use reacts useMemo

**Reply by Pelv** [reply]
https://beta.reactjs.org/apis/react/useMemo#should-you-add-usememo-everywhere

**Reply by yuvalshimoni** [reply]
It's not answered.. I trying to understand why memo with useState

**Reply by Pelv** [reply]
To ensure that the observer never change. 

UseMemo is useful for perf optimisation, it's not a "real memorisation". It allows you to skip some re-render, not all.

**Reply by ivanjeremic** [reply]
also when you pass a function to useState it is for more expensive initial state

**Reply by TkDodo** [reply]
> To ensure that the observer never change.

This is the correct answer. We want to create _one_ instance of the Observer for the lifetime of the component. I do have a separate article on why `useState` is the best way to achieve that:

https://tkdodo.eu/blog/use-state-for-one-time-initializations

**Comment by yuvalshimoni** [comment]
Can you explain a bit about notifyManager?

**Reply by TkDodo** [reply]
The `notifyManager` is a mechanism for us to batch certain updates together. If multiple calls would trigger an update in the same microtask, we generally defer the update to produce fewer re-renders.

**Comment by praveenkumar66** [comment]
Two months back I texted you in twitter asking about this. Now it is here. Thanks for your commitment. Great blog.

**Comment by samuellawrentz** [comment]
Yo thanks for making it look so simple!! This is an awesome library.

**Comment by ghost** [comment]
Thanks so much for the explanation. Btw. what program did you use to create the charts? They are beautiful and I'd like to create similar ones =)

**Reply by TkDodo** [reply]
It's called excalidraw and it's open source. You can use it here: https://excalidraw.com/

**Comment by j0j032** [comment]
Thank you Dominik! always good to be backstage 😎

**Comment by meteor91** [comment]
Hi! Can you explain how you force react to rerender, when data is loaded or when refetch called?

**Reply by TkDodo** [reply]
before v4, it was just a `forceUpdate`:

```
const [, forceUpdate] = React.useState(0)

forceUpdate(x => x + 1)
```

now, it's `useSyncExternalStore`, where you can implement a `subscribe` callback that will re-render your component and return a snapshot:

```
  React.useSyncExternalStore(
    React.useCallback(
      (onStoreChange) =>
        isRestoring
          ? () => undefined
          : observer.subscribe(notifyManager.batchCalls(onStoreChange)),
      [observer, isRestoring],
    ),
    () => observer.getCurrentResult(),
    () => observer.getCurrentResult(),
  )
```

**Comment by raphaeljcm** [comment]
That was a great article! After using RQ a lot for the past months, It is nice to understand how it works under the hood. Thanks a lot!

**Comment by StyleShit** [comment]
Great article! 🚀

So the `QueryObserver` is responsible for avoiding re-renders (or actually "not notifying") when the response is the same?
If so, can you please point me to the area in the code where it happens?

Thanks 🙏

**Reply by TkDodo** [reply]
it's in `shouldNotifyListeners` here: https://github.com/TanStack/query/blob/100ae76c402f862f4702eac30369340afd5b57c1/packages/query-core/src/queryObserver.ts#L634

**Reply by StyleShit** [reply]
I see that there is shallow comparison of the data, but doesn't the data reference always change on each refetch even if the response is identical? (e.g. you `JSON.parse()` the same string)

If so, how does this comparison work? I think I'm missing something here

**Reply by TkDodo** [reply]
`structuralSharing`, a "smart, deep-equal" copying, happens on Query level:

- https://github.com/TanStack/query/blob/792e0b24368b1650fdd904c9e4c85cdcf7118867/packages/query-core/src/utils.ts#L424-L437
- https://github.com/TanStack/query/blob/792e0b24368b1650fdd904c9e4c85cdcf7118867/packages/query-core/src/query.ts#L201

you can read more about it here: https://tkdodo.eu/blog/react-query-render-optimizations#structural-sharing

**Reply by StyleShit** [reply]
Oh, interesting!

So just to make sure I got it right, this is the flow?

1. Got new result from the `queryFn`
2. Create a smart deep copy of the result
3. Put it in the cache
4. Observer will run (and eventually call `queryObserver.updateResult( ... )`)
5. Shallowly compare the prev & next data (since we keep references using structural sharing for both cache & selectors)
    5.1. If the data has changed - notify listeners
    5.2. Otherwise - do nothing

**Reply by TkDodo** [reply]
pretty much, yeah. What the observer will do additionally is keeping track of which fields of the `result` is actually used by the component. So if you use `const { data, status } = useQuery(...)`, you will be notified of changes to the `data` and `status` field. If you also use `isFetching`, then your component will receive more updates, as this prop changes more often. This is know as "tracked queries", and is the default since v4: https://tkdodo.eu/blog/react-query-render-optimizations#tracked-queries

**Reply by StyleShit** [reply]
Just looked at the PR and it's really cool! (and smart with its simplicity) 

Thanks for all of your work, you're awesome 🙏

**Comment by woohm402** [comment]
Hi, thank you for writing a great article :)

What I understood is: `react-query` is just an adapter of `react` and `query-core`. When calling `useQuery`, QueryObserver is created, and then `query-core` subscribes / calls queryFn itself.

I found that when using `useQuery` without any options like `suspense`, the queryFn is triggered at the same level of `useEffect`, because this [`useEffect`](https://github.com/TanStack/query/blob/515803f5fa7ebc35ecb0504e60df9fe7e3bf057f/packages/react-query/src/useBaseQuery.ts#L97-L101) triggers `setOptions()` and this calls `queryFn`.

I think it could be on render phase instead of effect phase. Because useEffect is called after rendering is done, "triggered in useEffect" vs "triggered while rendering" will have a performance difference (yes, it will be small - about 10ms, though. 😅).

like this way (briefly writed)

```ts
// as is
// of course, indeed - here are many many more complex steps & options
const useQuery = (queryFn) => {
  useEffect(() => {
    queryFn();  // <- called while effect phase
  }, [queryFn]);
}

// to be
// something like this would be better in performance
const useQuery = (queryFn) => {
  const prevRef = useRef();

  if (someConditionCalculatedFromPrevRefCurrentValue) queryFn(); // <- called while rendering phase
}
```

This would be a terrible code in plain react, but react-query uses its own observer, so this pattern seems not that wierd.

So my question is: is this behavior intended despite this can lead to a little poor performance?

Again, thank you for a great article and library! I love tanstack query :)

**Reply by TkDodo** [reply]
render functions need to be pure. triggering a side effect like data fetching in them is technically against the rules of react. This can be troublesome when combined with `suspense` or concurrent features, where rendering can be interrupted. We also need an active observer first, and that subscription happens in `useSyncExternalStore`, which has similar timings to `useEffect`.

**Reply by woohm402** [reply]
Ah... I see. Thanks! 🙏

**Comment by trongtai37** [comment]
It's awesome!

**Comment by debater-coder** [comment]
How does it know which fields are being tracked?

**Reply by TkDodo** [reply]
We use Object.defineProperty with custom getters and track which field gets accessed on the result returned from useQuery

**Comment by mechaniccoder** [comment]
Thanks for the great article!

**Comment by idiglove** [comment]
I love it! Thanks for this. I have a question, how do you handle clean ups on useMutation? Do you use fetch API in the background, how do you handle abort controllers?

**Reply by TkDodo** [reply]
mutations can't be aborted because they have side-effects on the server

**Comment by dovranJorayev** [comment]
> I love it! Thanks for this. I have a question, how do you handle clean ups on useMutation? Do you use fetch API in the background, how do you handle abort controllers?

I guess you mean unsubcribe callbacks like in useEffect, then i guess generally it is unsafe to cancel mutations, and it is indeterminated what will happen on cancel, does data changed on the external storage or not. For this reason it is a [page in docs](https://tanstack.com/query/v4/docs/react/guides/query-cancellation) about query cancelation and is not about mutation cancelation 

And if you question about, how making a request happens inside useMutation, you are welcome to explore the source code of it 😄 
It is interesting part of open source i think

**Comment by mechaniccoder** [comment]
Great explanation of architecture!

**Comment by alexl-shopware** [comment]
Very well written and clear. I'm impressed. I'm a dev with 6 years experience but still have a hard time grasping architectures and technical details. You laid it out masterfully and was easy to follow along.

**Comment by morijenab** [comment]
Hi, 
Great blog thanks!
I have a question and hope here is the right place to ask. 

What is the main question:
Where is the React Query cache stored?

more context:
We are using micro front end structure more specifically( Single Spa package). 
Each applications has its own cache provider, when i switch between different applications the application is gone ( tested by `useeffect cleanup function` inside App level). 

What is strange to me?
when switching btw different applications: 
1-Application A mounted (react query cache is empty).  
2-Application B mounted and Application A unmounted. 
3-Application B unmounted and Application A mounted(now here before api calls finish the react query cache is exacllty like the last time it was mounted

Each application is a separate project being mounted/unmounted under the single spa management.

**Reply by TkDodo** [reply]
> Where is the React Query cache stored?

In memory, in the `queryClient`. When you do `new QueryClient()`, this is actually:

```
const queryClient = new QueryClient({
  queryCache: new QueryCache(),
  mutationCache: new MutationCache()
})
```

the client holds the cache, and the client lives in memory.

**Reply by morijenab** [reply]
Thanks for your response 

Lets consider this structure
this is the highest level in the project and as you can see the queryClient wrap the entire App.
if i switch the application i will see the console `unMounted`. here is the question what happen to the `QueryClientProvider` after this phase?
if it lives in the memory i expect this should be gone. however after i switch back to this app i can see the  `QueryClientProvider` has the last data. 

```js
export const Index = () => {
    useEffect(() => {
        return () => {
            console.log('unMounted!');
        };
    }, []);
    return (
        <CacheProvider value={customCache}>
            <ThemeProvider theme={defaultTheme}>
                <QueryClientProvider client={ReactQueryClient}>
                    <I18nextProvider i18n={i18n}>
                        <App />
                    </I18nextProvider>
                </QueryClientProvider>
            </ThemeProvider>
        </CacheProvider>
    );
};```

**Reply by TkDodo** [reply]
the provider is just react context. The question is where is the `ReactQueryClient` coming from...

**Comment by harsh-lamba** [comment]
Hi Great blogs, Thanks for the explanation.

But I have query along use case in our system and wanted your input in it:
Use Case:
We have hook (for instance useStatus) that wraps the react query. This hook is being used in parent child hierarchy . In parent component, this hook is being used and it is also being used in child and grand-child as well. But Grand child is being rendered asynchronously so we see two duplicate request on network. 

Reason behind: Why do we have duplicate request on Network tab?
Since components were rendered in different lifecycle, that is the reason we have two different network call. (Stale time:0 [Which is our requirement]). And we are facing these problems with a lot of components.

Following are the approaches that we thought to avoid these multiple request:
Use of refetchOnMount: false
This solution is doing background fetch when we come back from Route A to Route B (Already data was cached and stale).

Use of refetchOnMount: false + cacheTime: 1s
This solution will do background fetch but will also render the component (which we don't want).

May be we could put some staleTime: 1s:
With this we might be able to overcome with multiple duplicate request problem.

Also possibly we could avoid such instances by passing props from parent to child to grand child component but don't want to be a part of props drilling.

Could you provide your feedback for following:
-  if I understood the problem correctly for duplicate request?
-  is there any solution to avoid multiple duplicate refetch and also  refetch when user visit the page again?

React query version: 3.39.3
React: 17.0.2

**Reply by TkDodo** [reply]
My recommendation would be set a staleTime that is small enough to dedupe the requests in your scenario and leave everything else as default. Not sure why `staleTime:0` is your requirement? If it is, then you wouldn't be worried about duplicate requests, because that's the "feature" of it :)

If you set `staleTime` to e.g. 5 seconds, you will see your mounts on the same page that are in different lifecycles be deduplicated. And if the user navigates between pages, they will likely also get a background refetch. Sure, if they navigate between pages within 5 seconds, they won't see a refetch, but is that really a use-case?

**Comment by aiKhalil08** [comment]
Thanks for the article Dom. New to react qery and I have a question. Is the data in the query updated after EVERY fetch or ONLY when the data returned from the backend is new?

The diagram you used above seems to suggest the former. Thus the observer notifies the component after every refetch and causes a re render??

**Reply by TkDodo** [reply]
we put data in the cache after every fetch, which updates meta data like `dataUpdatedAt`. But the data we put in there is already [structurally shared](https://tanstack.com/query/v5/docs/framework/react/guides/render-optimizations#structural-sharing) with the previous data, so if nothing changed, it will be the exact same reference, thus not notifying observers.

**Comment by AtiqurRahmanAni** [comment]
How to remove inactive queries upon logout? I a page, I use two useQuery() to fetch data. I called queryClient.removeQueries(); and queryClient.cancelQueries(); when logged out. But in react-query-dev tool, still those two queries are present and it shows status is inactive.

**Reply by TkDodo** [reply]
show a reproduction please

**Reply by Evavic44** [reply]
```jsx
queryClient.clear() probably?
```
